plugins/modules/win_dns_zone.ps1

```diff
@@ -0,0 +1,231 @@
+                    Catch { $module.FailJson("Failed to convert DNS zone $($name): $($_.Exception.Message)", $_) }
+                }
+                Try { if ($dns_servers) { Set-DnsServerSecondaryZone -Name $name -MasterServers $dns_servers -WhatIf:$check_mode } }
+                Catch { $module.FailJson("Failed to set properties on the zone $($name): $($_.Exception.Message)", $_) }
+            }
+        }
+        "stub" {
+            $parms.Remove('DynamicUpdate')
+            if (-not $current_zone) {
+                # create zone
+                Try { Add-DnsServerStubZone @parms -WhatIf:$check_mode }
+                Catch { $module.FailJson("Failed to add $type zone $($name): $($_.Exception.Message)", $_) }
+            } else {
+                # update zone
+                if (-not $current_zone_type_match) { $module.FailJson("Failed to convert DNS zone $($name) to $type, unsupported conversion") }
+                Try {
+                    if ($parms.ReplicationScope) { Set-DnsServerStubZone -Name $name -ReplicationScope $parms.ReplicationScope -WhatIf:$check_mode }
+                    if ($forwarder_timeout) { Set-DnsServerStubZone -Name $name -ForwarderTimeout $forwarder_timeout -WhatIf:$check_mode }
+                    if ($dns_servers) { Set-DnsServerStubZone -Name $name -MasterServers $dns_servers -WhatIf:$check_mode }
+                }
+                Catch { $module.FailJson("Failed to set properties on the zone $($name): $($_.Exception.Message)", $_) }
+            }
+        }
+        "forwarder" {
+            # remove irrelevant params
+            $parms.Remove('DynamicUpdate')
+            $parms.Remove('ZoneFile')
+            if ($forwarder_timeout -and ($forwarder_timeout -in 0..15)) {
+                $parms.ForwarderTimeout = $forwarder_timeout
+            }
+            if ($forwarder_timeout -and -not ($forwarder_timeout -in 0..15)) {
+                $module.Warn("The forwarder_timeout param must be an integer value between 0 and 15")
+            }
+            if (-not $current_zone) {
+                # create zone
+                Try { Add-DnsServerConditionalForwarderZone @parms -WhatIf:$check_mode }
+                Catch { $module.FailJson("Failed to add $type zone $($name): $($_.Exception.Message)", $_) }
+            } else {
+                # update zone
+                if (-not $current_zone_type_match) { $module.FailJson("Failed to convert DNS zone $($name) to $type, unsupported conversion") }
+                Try {
+                    if ($parms.ReplicationScope) { Set-DnsServerConditionalForwarderZone -Name $name -ReplicationScope $parms.ReplicationScope -WhatIf:$check_mode }
+                    if ($forwarder_timeout) { Set-DnsServerConditionalForwarderZone -Name $name -ForwarderTimeout $forwarder_timeout -WhatIf:$check_mode }
+                    if ($dns_servers) { Set-DnsServerConditionalForwarderZone -Name $name -MasterServers $dns_servers -WhatIf:$check_mode }
```

Comment:
Have a look at using splatting for this instead of repeating the cmdlet call like

```suggestion
                    $set_params = @{}
                    if ($parms.ReplicationScope) { $set_params.ReplicationScope = $parms.ReplicationScope }
                    if ($forwarder_timeout) { $set_params.ForwarderTimeout = $forwarder_timeout }
                    if ($dns_servers) { $set_params.MasterServers = $dns_servers }
                    Set-DnsServerConditionalForwarderZone -Name $name -WhatIf:$check_mode @set_params
```

This is assuming you can set all those options in 1 call instead of separate ones though.

Reply:
I leverage this in other parts of the code but unfortunately with certain cmdlets (Set-DnsServerConditionalForwarderZone for example) you can't change more than one option in a single call - so I opted to take this approach. If there's a better approach i'm happy to explore it!